// redux-pender/lib/middleware.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.default = penderMiddleware;

var _actionTypes = require('./action-types');

var _actionTypes2 = _interopRequireDefault(_actionTypes);

var _penderize = require('./penderize');

var _penderize2 = _interopRequireDefault(_penderize);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var PENDING = _actionTypes2.default.PENDING,
    SUCCESS = _actionTypes2.default.SUCCESS,
    FAILURE = _actionTypes2.default.FAILURE;

/**
 * Middleware that pends and handles the promises
 * @param {object} config
 * @returns {function} middleware
 */

function penderMiddleware() {
    var config = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : { major: true };


    return function (store) {
        return function (next) {
            return function (action) {

                /**
                 * checks whether the given parameter is a promise
                 * 
                 * @param {object} promise 
                 * @returns {boolean} 
                 */
                function isPromise(promise) {
                    if (!promise) return false;
                    return promise.then && promise.catch;
                }

                /**
                 * extracts promise from the action
                 * 
                 * @param {object} action 
                 * @returns {Promise} promise
                 */
                function getPromise(action) {
                    var payload = action.payload;


                    if (!payload) return null; // there is no payload

                    // when 'major' option is true
                    if (config.major === true) {
                        if (isPromise(payload)) return payload;
                    }

                    // case when major is false
                    var pend = action.payload.pend;

                    if (isPromise(pend)) return pend;

                    return null;
                }

                var promise = getPromise(action); // get the promise from the action
                if (!promise) return next(action); // if there is no promise, skip this action


                var type = action.type,
                    meta = action.meta; // get the details of the action

                var penderized = (0, _penderize2.default)(type); // create the penderized action types


                // inform that the promise has started
                store.dispatch({
                    type: penderized.PENDING,
                    meta: meta
                });

                // log this in pender reducer
                store.dispatch({
                    type: PENDING,
                    payload: type
                });

                // handle the promise

                promise.then(function (result) {
                    // promise is resolved
                    // result will be assigned as payload
                    store.dispatch({
                        type: penderized.SUCCESS,
                        payload: result,
                        meta: meta
                    });

                    // log this in pender reducer
                    store.dispatch({
                        type: SUCCESS,
                        payload: type
                    });
                }).catch(function (error) {
                    // promise is rejected
                    // error will be assigned as payload
                    store.dispatch({
                        type: penderized.FAILURE,
                        payload: error,
                        meta: meta,
                        error: true
                    });

                    // log this in pender reducer
                    store.dispatch({
                        type: FAILURE,
                        payload: type
                    });
                });

                return promise;
            };
        };
    };
}
