// react-spinners/dist/spinners/ClimbingBoxLoader.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(['exports', 'react', 'prop-types', 'domkit/appendVendorPrefix', 'domkit/insertKeyframesRule'], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require('react'), require('prop-types'), require('domkit/appendVendorPrefix'), require('domkit/insertKeyframesRule'));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.react, global.propTypes, global.appendVendorPrefix, global.insertKeyframesRule);
    global.ClimbingBoxLoader = mod.exports;
  }
})(this, function (exports, _react, _propTypes, _appendVendorPrefix, _insertKeyframesRule) {
  'use strict';

  Object.defineProperty(exports, "__esModule", {
    value: true
  });

  var _react2 = _interopRequireDefault(_react);

  var _propTypes2 = _interopRequireDefault(_propTypes);

  var _appendVendorPrefix2 = _interopRequireDefault(_appendVendorPrefix);

  var _insertKeyframesRule2 = _interopRequireDefault(_insertKeyframesRule);

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
      default: obj
    };
  }

  function _classCallCheck(instance, Constructor) {
    if (!(instance instanceof Constructor)) {
      throw new TypeError("Cannot call a class as a function");
    }
  }

  var _createClass = function () {
    function defineProperties(target, props) {
      for (var i = 0; i < props.length; i++) {
        var descriptor = props[i];
        descriptor.enumerable = descriptor.enumerable || false;
        descriptor.configurable = true;
        if ("value" in descriptor) descriptor.writable = true;
        Object.defineProperty(target, descriptor.key, descriptor);
      }
    }

    return function (Constructor, protoProps, staticProps) {
      if (protoProps) defineProperties(Constructor.prototype, protoProps);
      if (staticProps) defineProperties(Constructor, staticProps);
      return Constructor;
    };
  }();

  function _possibleConstructorReturn(self, call) {
    if (!self) {
      throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
    }

    return call && (typeof call === "object" || typeof call === "function") ? call : self;
  }

  function _inherits(subClass, superClass) {
    if (typeof superClass !== "function" && superClass !== null) {
      throw new TypeError("Super expression must either be null or a function, not " + typeof superClass);
    }

    subClass.prototype = Object.create(superClass && superClass.prototype, {
      constructor: {
        value: subClass,
        enumerable: false,
        writable: true,
        configurable: true
      }
    });
    if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass;
  }

  /**
   * @type {object}
   */
  var keyframes = {
    '0%': {
      transform: 'translate(0, -1em) rotate(-45deg)'
    },
    '5%': {
      transform: 'translate(0, -1em) rotate(-50deg)'
    },
    '20%': {
      transform: 'translate(1em, -2em) rotate(47deg)'
    },
    '25%': {
      transform: 'translate(1em, -2em) rotate(45deg)'
    },
    '30%': {
      transform: 'translate(1em, -2em) rotate(40deg)'
    },
    '45%': {
      transform: 'translate(2em, -3em) rotate(137deg)'
    },
    '50%': {
      transform: 'translate(2em, -3em) rotate(135deg)'
    },
    '55%': {
      transform: 'translate(2em, -3em) rotate(130deg)'
    },
    '70%': {
      transform: 'translate(3em, -4em) rotate(217deg)'
    },
    '75%': {
      transform: 'translate(3em, -4em) rotate(220deg)'
    },
    '100%': {
      transform: 'translate(0, -1em) rotate(-225deg)'
    }
  };

  /**
   * @type {string}
   */
  var animationName = (0, _insertKeyframesRule2.default)(keyframes);

  var Loader = function (_React$Component) {
    _inherits(Loader, _React$Component);

    function Loader() {
      _classCallCheck(this, Loader);

      return _possibleConstructorReturn(this, (Loader.__proto__ || Object.getPrototypeOf(Loader)).apply(this, arguments));
    }

    _createClass(Loader, [{
      key: 'getBoxStyle',
      value: function getBoxStyle() {
        return {
          position: 'absolute',
          left: '0px',
          bottom: '-0.1em',
          width: '1em',
          height: '1em',
          backgroundColor: 'transparent',
          border: '0.25em solid ' + this.props.color,
          borderRadius: '15%',
          transform: 'translate(0, -1em) rotate(-45deg)'
        };
      }
    }, {
      key: 'getHillStyle',
      value: function getHillStyle() {
        return {
          position: 'absolute',
          width: '7.1em',
          height: '7.1em',
          top: '1.7em',
          left: '1.7em',
          borderLeft: '0.25em solid ' + this.props.color,
          transform: 'rotate(45deg)'
        };
      }
    }, {
      key: 'getAnimationStyle',
      value: function getAnimationStyle() {
        var animation = [animationName, '2.5s', 'infinite', 'cubic-bezier(.79, 0, .47, .97)'].join(' ');
        var animationFillMode = 'both';

        return {
          animation: animation,
          animationFillMode: animationFillMode
        };
      }
    }, {
      key: 'getStyle',
      value: function getStyle() {
        return (0, _appendVendorPrefix2.default)(this.getBoxStyle(), this.getAnimationStyle());
      }
    }, {
      key: 'renderLoader',
      value: function renderLoader(loading) {
        if (loading) {
          var style = {
            position: 'absolute',
            top: '50%',
            left: '50%',
            marginTop: '-2.7em',
            marginLeft: '-2.7em',
            width: '5.4em',
            height: '5.4em',
            fontSize: this.props.size
          };

          return _react2.default.createElement(
            'div',
            { className: 'react-spinners--climbing-box' },
            _react2.default.createElement(
              'div',
              { style: style },
              _react2.default.createElement('div', { style: this.getStyle() }),
              _react2.default.createElement('div', { style: this.getHillStyle() })
            )
          );
        }

        return null;
      }
    }, {
      key: 'render',
      value: function render() {
        return this.renderLoader(this.props.loading);
      }
    }]);

    return Loader;
  }(_react2.default.Component);

  /**
   * @type {object}
   */
  Loader.propTypes = {
    loading: _propTypes2.default.bool,
    color: _propTypes2.default.string,
    size: _propTypes2.default.number
  };

  /**
   * @type {object}
   */
  Loader.defaultProps = {
    loading: true,
    color: '#000000',
    size: 15
  };

  exports.default = Loader;
});
